### PR TITLE
validate crypto command

### LINKED ISSUE #388: Invalid certificate after adding/removing core ASes to ISD
On 3.09.2021 we attempted to remove some core AS from ISD 19 and add others, as core, to it.
There was a problem and although the TRC was newly generated, the certificates couldn't be validated with them, causing the `CS` to fail at start up.
The steps to reproduce the problem are:
* Delete or make any change to one of the core ASes.
* The ISD regenerates the core ASes' certificates and the TRC 
* When we write the certificates, _eventually_ we will hit the problem of writing the old one overwriting the new one (because we get the `certificates()` without order

Fix the bug and fix the topology.

### PR DESCRIPTION
**Pull request by juagargi**
- [x] add a `validate` method to the AS model
- [x] add corresponding UTs

Closes #388

---
This change is [Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)

### CODE REVIEW

**Review by matzf — CHANGES_REQUESTED** [review]
> Reviewed 1 of 13 files at r1.
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: 1 of 13 files reviewed, 3 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/hostfiles/scionlab-config, line 197 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-rVXF6ePPpA645_ac:-Mj-rVXF6ePPpA645_ad:b-89ipkp) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/hostfiles/scionlab-config#L197)):*
> ```
>     withfile_or_noinstall.add_argument(
>         '--noinstall',
>         # type=argparse.FileType('wb'),
> ```

Left over comment.
___
*[scionlab/hostfiles/scionlab-config, line 259 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-r-1f-xftcUr0yMLG:-Mj-r-1f-xftcUr0yMLH:b-rkllhs) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/hostfiles/scionlab-config#L259)):*
> ```
>         fetch_info = fetch_info._replace(url=args.url)
> 
>     if args.force:
> ```

` or args.noinstall`, as we want to unconditionally fetch the tar file.
___
*[scionlab/hostfiles/scionlab-config, line 338 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-pz5g8dSLY6tRkBSG:-Mj-pz5g8dSLY6tRkBSH:b-300hi4) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/hostfiles/scionlab-config#L338)):*
> ```
> 
> def write_tarfile(filepath, tarbytes):
>     if isinstance(tarbytes, bytes):
> ```

The only case that would fail this check is empty configuration. Perhaps this would be more explicit as `tarbytes is not _CONFIG_EMPTY`.  What do we do in the empty case? Maybe log.warn?

**Review by matzf — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: 1 of 13 files reviewed, 10 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/management/commands/check_crypto.py, line 42 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-xgAf2W9gDmuZJ_U0:-Mj-xgAf2W9gDmuZJ_U1:b-d908pg) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/management/commands/check_crypto.py#L42)):*
> ```Python
>                 try:
>                     num = as_.validate_crypto()
>                     if (as_.is_core and num != 5) or (not as_.is_core and num != 1):
> ```

Why 5? Couldn't this check be moved into `validate_crypto`?
___
*[scionlab/scion/certs.py, line 67 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-uAg1CuaNUnm_Btrb:-Mj-uAg2DNjhj1_Va4Vm:bq21rm1) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L67)):*
> ```Python
> 
> 
> def verify_certificate_valid(cert: bytes, cert_usage: str):
> ```

Nit, it would seem to be more consistent if these verify functions take the cert/key data as "pem" string instead of as a byte sequence, like the `decode_certificate` and `decode_key` functions.
___
*[scionlab/scion/certs.py, line 86 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-wGpCD1kbQRfjovpt:-Mj-wGpCD1kbQRfjovpu:b5qtxmu) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L86)):*
> ```Python
>     with contextlib.ExitStack() as stack:
>         trc_file = stack.enter_context(NamedTemporaryFile(suffix=".trc"))
>         cert_file = stack.enter_context(NamedTemporaryFile(suffix=".pem"))
> ```

The ExitStack thing is only needed if there is a dynamic number of things. This can be:
`with NamedTemporaryFile(suffix=".trc") as trc_file, NamedTemporaryFile(suffix=".pem") as cert_file:`
___
*[scionlab/scion/certs.py, line 287 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-wrkV-1SDPRd9dB-O:-Mj-wrkV-1SDPRd9dB-P:b-6xibla) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L287)):*
> ```Python
> 
> 
> def _run_scion_pki(*args, cwd=None, check=True):
> ```

Nit, call this `_run_scion_pki_certificate`.
___
*[scionlab/scion/keys.py, line 68 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-wvma52rCtc8IwGUZ:-Mj-wvma52rCtc8IwGU_:bl1xsii) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/keys.py#L68)):*
> ```Python
> 
> 
> def _run_scion_pki(*args, cwd=None, check=True):
> ```

Nit, call this `_run_scion_pki_key`.
___
*[scionlab/scion/trcs.py, line 258 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-tFLK6gFoW4UI9qwH:-Mj-tFLK6gFoW4UI9qwI:bhu9d1c) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/trcs.py#L258)):*
> ```Python
> 
> 
> def _run_scion_pki(*args, cwd=None, check=True):
> ```

Nit, call this `_run_scion_pki_trcs`?
___
*[scionlab/scion/util.py, line 16 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-sACM7ZzUe2DZT5LA:-Mj-sACM7ZzUe2DZT5LB:b54u5hk) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/util.py#L16)):*
> ```Python
> 
> """
> :mod:`scionlab.scion.util` --- scion-pki execution
> ```

Nit, could we give the module a more specific name? `pkicommand.py` or something.

**Review by juagargi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: 1 of 13 files reviewed, 10 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/hostfiles/scionlab-config, line 197 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-rVXF6ePPpA645_ac:-Mj-saLY1o8UcKQlPHOl:b-896fix) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/hostfiles/scionlab-config#L197)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Left over comment.
</blockquote></details>

Done.
___
*[scionlab/hostfiles/scionlab-config, line 259 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-r-1f-xftcUr0yMLG:-Mj-smqzCFMxBwRvjqW4:b-896fix) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/hostfiles/scionlab-config#L259)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

` or args.noinstall`, as we want to unconditionally fetch the tar file.
</blockquote></details>

Done.
___
*[scionlab/hostfiles/scionlab-config, line 338 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-pz5g8dSLY6tRkBSG:-Mj-ydt33xZmqgO9ZUv6:bc56g2y) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/hostfiles/scionlab-config#L338)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

The only case that would fail this check is empty configuration. Perhaps this would be more explicit as `tarbytes is not _CONFIG_EMPTY`.  What do we do in the empty case? Maybe log.warn?
</blockquote></details>

I have decided for both the warning and creating an empty file (in case we rely on the existence of a valid tar file)

**Review by juagargi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: 1 of 13 files reviewed, 10 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/management/commands/check_crypto.py, line 42 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-xgAf2W9gDmuZJ_U0:-Mj37tK-5OGwwLGMazIo:b17dj1h) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/management/commands/check_crypto.py#L42)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Why 5? Couldn't this check be moved into `validate_crypto`?
</blockquote></details>

It could be moved there.

**Review by juagargi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: 1 of 13 files reviewed, 6 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/scion/certs.py, line 67 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-uAg1CuaNUnm_Btrb:-MjZ7G6CDcdCT0of7AMQ:b-o7oe0b) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L67)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Nit, it would seem to be more consistent if these verify functions take the cert/key data as "pem" string instead of as a byte sequence, like the `decode_certificate` and `decode_key` functions.
</blockquote></details>

All the validate functions take the deserialized object (TRC, cert or key). It's just that it's been represented as bytes instead of a fully blown object (x509 or EC). The data flow is: 1st obtain stream and deserialize it, and 2nd: manipulate object (TRC, cert or key).
The representation of a certificate or a key is as bytes, since the tooling underneath (`scion-pki`) accepts only bytes.
___
*[scionlab/scion/certs.py, line 86 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-wGpCD1kbQRfjovpt:-MjZ3pwr2JbNvDSAd4o9:b-896fix) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L86)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

The ExitStack thing is only needed if there is a dynamic number of things. This can be:
`with NamedTemporaryFile(suffix=".trc") as trc_file, NamedTemporaryFile(suffix=".pem") as cert_file:`
</blockquote></details>

Done.
___
*[scionlab/scion/certs.py, line 287 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-wrkV-1SDPRd9dB-O:-MjZ4BTU7UHJzHNxNC1u:b-896fix) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L287)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Nit, call this `_run_scion_pki_certificate`.
</blockquote></details>

Done.
___
*[scionlab/scion/keys.py, line 68 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-wvma52rCtc8IwGUZ:-MjZ4Oad9-89DwUgnamv:b-896fix) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/keys.py#L68)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Nit, call this `_run_scion_pki_key`.
</blockquote></details>

Done.
___
*[scionlab/scion/trcs.py, line 258 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-tFLK6gFoW4UI9qwH:-MjZ4_9t50LgJXn3g4fz:b-896fix) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/trcs.py#L258)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Nit, call this `_run_scion_pki_trcs`?
</blockquote></details>

Done.
___
*[scionlab/scion/util.py, line 16 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-sACM7ZzUe2DZT5LA:-MjZ65SI4eRHhrqkHUkA:b-7rl0jd) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/util.py#L16)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Nit, could we give the module a more specific name? `pkicommand.py` or something.
</blockquote></details>

done.

**Review by matzf — COMMENTED** [review]
> Reviewed 4 of 13 files at r1, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: all files reviewed, 2 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/scion/certs.py, line 67 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-uAg1CuaNUnm_Btrb:-MjZuLzT2MdiG9XU6h9t:bf3hrkm) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L67)):*
<details><summary><i>Previously, juagargi (Juan A. Garcia Pardo) wrote…</i></summary><blockquote>

All the validate functions take the deserialized object (TRC, cert or key). It's just that it's been represented as bytes instead of a fully blown object (x509 or EC). The data flow is: 1st obtain stream and deserialize it, and 2nd: manipulate object (TRC, cert or key).
The representation of a certificate or a key is as bytes, since the tooling underneath (`scion-pki`) accepts only bytes.
</blockquote></details>

Sorry, I meant consistent with e.g. `decode_certificate` which takes the thing as a "pem" string (as this is how it's stored in the DB). Then all the `encode/decode("ascii")` are in the same place.

**Review by juagargi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: all files reviewed, 2 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/scion/certs.py, line 67 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-uAg1CuaNUnm_Btrb:-MjcJkRYAF3oRAqbLSu2:bmuiqwz) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L67)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

Sorry, I meant consistent with e.g. `decode_certificate` which takes the thing as a "pem" string (as this is how it's stored in the DB). Then all the `encode/decode("ascii")` are in the same place.
</blockquote></details>

TL;DR see "note 1" below.

Yes, that is what I understood, and the reason why I explained above about the signatures of all `verify_` functions. Let me explain it with a current usage of `verify_trcs`: in the function `tests.utils.check_trc` 
https://github.com/netsec-ethz/scionlab/blob/juagargi/validate_crypto/scionlab/tests/utils.py#L267
, in particular in this section of code:
```python
    trc_pld = trcs.decode_trc(trc.trc)
...
    trcs.verify_trcs(prev_trc_pld, trc_pld)
```
we have to do several things with a TRC, namely convert it to a dictionary via `trcs.trc_to_dict(trc_pld)` and also verify the TRC chain via `trcs.verify_trcs(prev_trc_pld, trc_pld)`. That is why the functions accept the TRCs _as deserialized objects_, not as their DB representation. The DB representation does not affect these functions that manipulate objects.
The way to proceed is: 1) obtain the data from the DB and convert it to its manipulable representation (`bytes` for TRCs,  and`cryptography` objects for certificates and keys), and 2) then manipulate the object, and 3) if we need to store again the object, serialize it back to its DB representation and store it.
Otherwise, all functions that manipulate the objects would have to deserialize the streams in each of their calls, which is wrong.
The coupled calls `encode_*` and `decode_*` _should_ be the only ones that can see the objects as both serialized for DB and manipulable objects, and thus the only calls that have `str` in their signatures.
Note 1: the `verify_*` functions for certificates and keys take `bytes` (instead of their proper `cryptography` objects) as a shortcut, as the functions (for the moment) only need to write to disk. This might be the source of the confusion. They _should_ take the `cryptography` objects, but I thought that for the moment it was a bit of a waste of processing.

I hope I explained my view more clearly with the current example that we have in `check_trc`.
In any case, let me know if this is really a nit, or it is something more important for you, and then we can decide how to move forward.

**Review by matzf — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: all files reviewed, 2 unresolved discussions / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/scion/certs.py, line 67 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-uAg1CuaNUnm_Btrb:-MjcUlx6-Bs7wzrfhsdg:b4xpujg) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L67)):*
<details><summary><i>Previously, juagargi (Juan A. Garcia Pardo) wrote…</i></summary><blockquote>

TL;DR see "note 1" below.

Yes, that is what I understood, and the reason why I explained above about the signatures of all `verify_` functions. Let me explain it with a current usage of `verify_trcs`: in the function `tests.utils.check_trc` 
https://github.com/netsec-ethz/scionlab/blob/juagargi/validate_crypto/scionlab/tests/utils.py#L267
, in particular in this section of code:
```python
    trc_pld = trcs.decode_trc(trc.trc)
...
    trcs.verify_trcs(prev_trc_pld, trc_pld)
```
we have to do several things with a TRC, namely convert it to a dictionary via `trcs.trc_to_dict(trc_pld)` and also verify the TRC chain via `trcs.verify_trcs(prev_trc_pld, trc_pld)`. That is why the functions accept the TRCs _as deserialized objects_, not as their DB representation. The DB representation does not affect these functions that manipulate objects.
The way to proceed is: 1) obtain the data from the DB and convert it to its manipulable representation (`bytes` for TRCs,  and`cryptography` objects for certificates and keys), and 2) then manipulate the object, and 3) if we need to store again the object, serialize it back to its DB representation and store it.
Otherwise, all functions that manipulate the objects would have to deserialize the streams in each of their calls, which is wrong.
The coupled calls `encode_*` and `decode_*` _should_ be the only ones that can see the objects as both serialized for DB and manipulable objects, and thus the only calls that have `str` in their signatures.
Note 1: the `verify_*` functions for certificates and keys take `bytes` (instead of their proper `cryptography` objects) as a shortcut, as the functions (for the moment) only need to write to disk. This might be the source of the confusion. They _should_ take the `cryptography` objects, but I thought that for the moment it was a bit of a waste of processing.

I hope I explained my view more clearly with the current example that we have in `check_trc`.
In any case, let me know if this is really a nit, or it is something more important for you, and then we can decide how to move forward.
</blockquote></details>

I see, thanks for the extended explanation. I did not remember that the DB representation of the TRCs is an arbitrary format that we chose. For the keys and certs, however, the DB format is already the proper PEM format. Explicitly `encode`ing the string appears unnecessary (and confusing, to me) here -- if we just pass these certs/keys as the PEM strings that they already are that doesn't change anything in the writing to disk (just open the named temporary file in text mode instead of binary mode) and gets rid of three `encode('ascii')` calls in `validate_crypto`.

No it's fine, this is not blocking.

**Review by juagargi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390)* status: 7 of 13 files reviewed, 1 unresolved discussion / 0 of 1 LGTMs obtained / 0 of 1 approvals obtained

___
*[scionlab/management/commands/check_crypto.py, line 42 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-xgAf2W9gDmuZJ_U0:-MjcjQ3m8mbh4A1XgkCo:b-896fix) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/management/commands/check_crypto.py#L42)):*
<details><summary><i>Previously, juagargi (Juan A. Garcia Pardo) wrote…</i></summary><blockquote>

It could be moved there.
</blockquote></details>

Done.
___
*[scionlab/scion/certs.py, line 67 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-uAg1CuaNUnm_Btrb:-MjcjvLn3luFhV0jkzle:b725pul) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L67)):*
<details><summary><i>Previously, matzf (Matthias Frei) wrote…</i></summary><blockquote>

I see, thanks for the extended explanation. I did not remember that the DB representation of the TRCs is an arbitrary format that we chose. For the keys and certs, however, the DB format is already the proper PEM format. Explicitly `encode`ing the string appears unnecessary (and confusing, to me) here -- if we just pass these certs/keys as the PEM strings that they already are that doesn't change anything in the writing to disk (just open the named temporary file in text mode instead of binary mode) and gets rid of three `encode('ascii')` calls in `validate_crypto`.

No it's fine, this is not blocking. 
</blockquote></details>

Updated (although it goes against design principles :smile: and now we have inconsistent inconsistencies)

**Review by matzf — APPROVED** [review]
:lgtm_strong:

Thanks!

> Reviewed 6 of 6 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-:-Mjclhzw3Pl8vq4lpyXz:b-afugb9)* status: :shipit: complete! all files reviewed, all discussions resolved

___
*[scionlab/scion/certs.py, line 67 at r1](https://reviewable.io/reviews/netsec-ethz/scionlab/390#-Mj-uAg1CuaNUnm_Btrb:-Mjcky0s3IKRAFJrVCZg:b-y59xdg) ([raw file](https://github.com/netsec-ethz/scionlab/blob/430a0d0b04e2124d0df0cbc90356df575864d48b/scionlab/scion/certs.py#L67)):*
<details><summary><i>Previously, juagargi (Juan A. Garcia Pardo) wrote…</i></summary><blockquote>

Updated (although it goes against design principles :smile: and now we have inconsistent inconsistencies)
</blockquote></details>

Thanks :+1:
